### PR TITLE
#447 save instruction statistics

### PR DESCRIPTION
**Pull request by otselnik**

### CODE REVIEW

**Comment by otselnik** [comment]
Move step count to solana_neon_transactions
Add heap info

**Review by afalaleev — APPROVED** [review]

File: proxy/indexer/accounts_db.py [hunk 1]
```diff
@@ -30,17 +30,28 @@ def set_acc_by_request(self, neon_account: str, pda_account: str, code_account:
                 ''',
                 (neon_account, pda_account, code_account, 0, code))
 
-    def set_acc_indexer(self, neon_account: str, pda_account: str, code_account: str, slot: int):
+    def set_acc_indexer(self, neon_account: str, pda_account: str, code_account: str, slot: int, sol_sign: str):
+        if not neon_account:
```
Comment (rozhkovdmitrii) [inline proxy/indexer/accounts_db.py]: does it pretend to be a separate method?

File: proxy/db/scheme.sql [hunk 2]
```diff
@@ -123,6 +125,17 @@
...
 
+    CREATE TABLE IF NOT EXISTS solana_neon_transactions_costs (
+        sol_sign CHAR(88) UNIQUE,
+        operator VARCHAR(50),
+
+        heap_size INT,
+        bpf_instructions INT,
+
+        sol_cost BIGINT,
+        token_income BIGINT
```
Comment (rozhkovdmitrii) [inline proxy/db/scheme.sql]: `neon_income`

File: proxy/db/scheme.sql [hunk 3]
```diff
@@ -26,6 +26,7 @@
         code_account VARCHAR(50),
         slot BIGINT,
         code TEXT,
+        sol_sign CHAR(88),
```
Comment (rozhkovdmitrii) [inline proxy/db/scheme.sql:29]: the first field called same as the table?

`neon_account` - `neon_accounts`

can we rename the key?

File: proxy/indexer/accounts_db.py [hunk 4]
```diff
@@ -30,17 +30,28 @@ def set_acc_by_request(self, neon_account: str, pda_account: str, code_account:
                 ''',
                 (neon_account, pda_account, code_account, 0, code))
 
-    def set_acc_indexer(self, neon_account: str, pda_account: str, code_account: str, slot: int):
+    def set_acc_indexer(self, neon_account: str, pda_account: str, code_account: str, slot: int, sol_sign: str):
```
Comment (rozhkovdmitrii) [inline proxy/indexer/accounts_db.py]:
```
def set_acc_indexer(self, neon_account: NeonAccount)
```

File: proxy/indexer/costs_db.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+from typing import List
+
+from ..indexer.utils import CostInfo
```
Comment (rozhkovdmitrii) [inline proxy/indexer/costs_db.py:3]:
```py
from ..indexer.data import CostInfo
```

File: proxy/indexer/costs_db.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class CostsDB(BaseDB):
+    def __init__(self):
+        BaseDB.__init__(self, 'solana_neon_transactions_costs')
+
+    def add_costs(self, tx_costs: List[CostInfo]):
+        rows = []
+        for cost_info in tx_costs:
+            rows.append((
```
Comment (rozhkovdmitrii) [inline proxy/indexer/costs_db.py]: is it applicable like that:

`rows.append(*const) ` or `rows.append(**costs)`

File: proxy/indexer/costs_db.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+                cost_info.heap,
+                cost_info.bpf,
+
+                cost_info.sol_spent,
+                cost_info.token_income
+            ))
+
+        with self._conn.cursor() as cursor:
+            cursor.executemany(f'''
+                INSERT INTO {self._table_name}
```
Comment (rozhkovdmitrii) [inline proxy/indexer/costs_db.py]: get rid of table_name

File: proxy/indexer/utils.py [hunk 8]
```diff
@@ -38,6 +39,63 @@ def __eq__(self, other):
     def get_req_id(self):
         return f"{self.idx}{self.sign}"[:7]
 
+    def set_steps(self, steps: int):
+        self.steps = steps
+
+
+class CostInfo:
+    def __init__(self, sign: str, tx: dict, program: PublicKey):
```
Comment (rozhkovdmitrii) [inline proxy/indexer/utils.py]: from typing import Dict
tx: Dict[str, Any]

File: proxy/indexer/utils.py [hunk 9]
```diff
@@ -38,6 +39,63 @@ def __eq__(self, other):
...
+        self.sign = sign
+        self.operator = None
+        self.sol_spent = None
+        self.bpf = None
+        self.heap = None
+        self.token_income = None
+        if tx:
+            self.setup(tx, program)
```
Comment (rozhkovdmitrii) [inline proxy/indexer/utils.py]: setup is bad

`init_from_tx`

File: proxy/indexer/utils.py [hunk 10]
```diff
@@ -38,6 +39,63 @@ def __eq__(self, other):
...
+            if len(msg_words) >= 7 and\
+                msg_words[0] == 'Program' and\
+                msg_words[1] == str(program) and\
+                msg_words[2] == 'consumed' and\
+                msg_words[4] == 'of' and\
+                msg_words[6] == 'compute' and\
+                msg_words[7] == 'units':
+                return int(log_words[3])
+            return None
+        def heap_log(msg_words: list[str]):
```
Comment (rozhkovdmitrii) [inline proxy/indexer/utils.py]: Send me please the example and I'll help you with it

File: proxy/indexer/utils.py [hunk 11]
```diff
@@ -38,6 +39,63 @@ def __eq__(self, other):
...
+        self.operator = None
+        self.sol_spent = None
+        self.bpf = None
+        self.heap = None
+        self.token_income = None
+        if tx:
+            self.setup(tx, program)
+
+    def setup(self, tx: dict, program: PublicKey):
+        def bpf_log(program, msg_words: list[str]):
```
Comment (rozhkovdmitrii) [inline proxy/indexer/utils.py]: move it just after the current method

File: proxy/indexer/utils.py [hunk 12]
```diff
@@ -38,6 +39,63 @@ def __eq__(self, other):
...
+                msg_words[1] == "log:" and\
+                msg_words[2] == 'Total' and\
+                msg_words[3] == 'memory' and\
+                msg_words[4] == 'occupied:':
+                return int(log_words[5])
+            return None
+
+        self.operator = tx['transaction']['message']['accountKeys'][0]
+        self.sol_spent = tx['meta']['preBalances'][0] - tx['meta']['postBalances'][0]
+        for log in tx['meta']['logMessages']:
```
Comment (rozhkovdmitrii) [inline proxy/indexer/utils.py]: may be it pretends to be simplified

**Review by rozhkovdmitrii — CHANGES_REQUESTED** [review]

File: proxy/indexer/accounts_db.py [hunk 13]
```diff
@@ -1,46 +1,71 @@
...
-        self.pda_account = pda_account
-        self.code_account = code_account
+    neon_address: str = None
+    pda_address: str = None
+    code_address: str = None
+    slot: int = 0
+    code: str = None
+    sol_sign: str = None
+
+    def __init__(self, neon_address: str = None, pda_address: str = None, code_address: str = None, slot: int = 0, code: str = None, sol_sign: str = None):
```
Comment (rozhkovdmitrii) [inline proxy/indexer/accounts_db.py]: The constructor is supplied by default by the `@dataclass`. You can easily get rid of it

File: proxy/indexer/accounts_db.py [hunk 14]
```diff
@@ -1,46 +1,71 @@
...
+                VALUES(%s, %s, %s, %s, %s, %s)
+                ON CONFLICT (pda_address, code_address) DO UPDATE
                 SET
                     slot=EXCLUDED.slot
                 ;
                 ''',
-                (neon_account, pda_account, code_account, slot))
+                tuple(neon_account))
+
+    def fill_neon_address_if_missing(self, neon_account: NeonAccountInfo) -> NeonAccountInfo:
```
Comment (rozhkovdmitrii) [inline proxy/indexer/accounts_db.py]: def fill_neon_address_if_missing(self, neon_account: NeonAccountInfo) -> bool:

File: proxy/indexer/accounts_db.py [hunk 15]
```diff
@@ -1,46 +1,71 @@
...
+        if not neon_account.neon_address:
```
Comment (rozhkovdmitrii) [inline proxy/indexer/accounts_db.py]: Just use `not None` to check it for `None`

**Review by rozhkovdmitrii — CHANGES_REQUESTED** [review]

**Review by rozhkovdmitrii — APPROVED** [review]

**Pull request merged**